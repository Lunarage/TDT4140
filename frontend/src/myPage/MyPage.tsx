import { useEffect, useState } from "react";
import { useDispatch, useSelector } from 'react-redux';
import styled from "styled-components";
import {
  PageHeader as BasePageHeader,
  RightWrapper,
  FilterWrapper,
  PageWrapper,
} from "../browse/Browse";
import ActivityDashboard from '../components/ActivityDashboard';
import Header from "../components/Header";
import { HeaderItem, HeaderItemUnderlined } from '../components/NewActivity';
import { getMyActivities, getStarred, getSignUps } from '../store/actionCreators';
import { State } from '../store/types';
import { Event } from '../store/types';
import { redHexColor } from '../consts';

const PageHeader = styled(BasePageHeader)`
  text-align: center;
`;

const TabsWrapper = styled(FilterWrapper)`
  display: flex;
  justify-content: space-evenly;
  align-items: center;
`;

const Tab = styled(HeaderItem)`
  color: ${redHexColor};
  font-size: 1.5em;
`;

const TabUnderlined = styled(HeaderItemUnderlined)`
  color: ${redHexColor};
  font-size: 1.5em;
`;

enum tab {
  starred = "Favoritter",
  signUp = "Oppmeldte aktiviteter",
  myAct = "Mine aktiviteter",
}

const MyPage = () => {
  const dispatch = useDispatch();

  const [currentTab, setCurrentTab] = useState<tab>(tab.myAct)
  const [tabComponents, setTabComponents] = useState<JSX.Element[]>()

  const [events, setEvents] = useState<Event[] | undefined>()
  const [isLoading, setIsLoading] = useState<boolean>(false)
  const [error, setError] = useState<any>()

  const {
    starred,
    isLoading: starredIsLoading,
    errorMessage: starredError,
  } = useSelector((state: State) => state.starredReducer);

  const {
    signUps,
    isLoading: signUpsIsLoading,
    errorMessage: signUpsError,
  } = useSelector((state: State) => state.signUpsReducer);

  const {
    myActivities,
    isLoading: myActivitiesIsLoading,
    errorMessage: myActivitiesError,
  } = useSelector((state: State) => state.myActivitiesReducer);

  useEffect(() => {
    const token = localStorage.getItem("token")
    const id = localStorage.getItem("id")
    if (id && token) {
      if (!starred && !starredIsLoading) {
        dispatch(getStarred(id, token))
      }
      if (!signUps && !signUpsIsLoading) {
        dispatch(getSignUps(id, token))
      }
      if (!myActivities && !myActivitiesIsLoading) {
        dispatch(getMyActivities(id, token))
      }
    }
    changeTab(currentTab)
  }, [starred, myActivities, signUps, currentTab]);

  const changeTab = (newTab: tab) => {
    setCurrentTab(newTab);
    if (newTab == tab.starred) {
      setEvents(starred)
      setIsLoading(starredIsLoading)
      setError(starredError)
    }
    else if (newTab == tab.signUp) {
      setEvents(signUps)
      setIsLoading(signUpsIsLoading)
      setError(signUpsError)
    }
    else if (newTab == tab.myAct) {
      setEvents(myActivities)
      setIsLoading(myActivitiesIsLoading)
      setError(myActivitiesError)
    }
  }

  useEffect(() => {
    const tabList = [tab.myAct, tab.signUp, tab.starred]
    let tempTabComponents: JSX.Element[] = []
    tabList.forEach(t => {
      if (t == currentTab) {
        tempTabComponents.push(<TabUnderlined onClick={() => changeTab(t)}>{t}</TabUnderlined>)
      } else {
        tempTabComponents.push(<Tab onClick={() => changeTab(t)}>{t}</Tab>)
      }
    })
    setTabComponents(tempTabComponents)
  }, [currentTab])

  return (
    <PageWrapper>
      <Header />
      <RightWrapper>
        <PageHeader>Min side</PageHeader>
        <TabsWrapper>
          {tabComponents}
        </TabsWrapper>
        <ActivityDashboard events={events} isLoading={isLoading} error={error} />
      </RightWrapper>
    </PageWrapper>
  );
};

export default MyPage;
