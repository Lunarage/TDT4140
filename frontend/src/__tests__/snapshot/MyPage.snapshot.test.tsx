import { mount } from "enzyme";
import MyPage from '../../myPage/MyPage';
import { store } from '../..';
import Enzyme from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJSON from "enzyme-to-json";
import { Provider } from "react-redux";

Enzyme.configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  ...jest.requireActual('react-router-dom') as any,
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

const MyPageWrapper = () => (
  <Provider store={store}>
    <MyPage />
  </Provider>
);

describe("MyPage", () => {
  it("renders MyPage", () => {
    mount(<MyPageWrapper />);
  });
  it("Should match MyPage snapshot", () => {
    const comp = mount(<MyPageWrapper />);
    expect(toJSON(comp)).toMatchSnapshot();
  });
});
