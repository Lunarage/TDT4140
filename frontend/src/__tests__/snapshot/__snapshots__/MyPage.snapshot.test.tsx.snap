// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyPage Should match MyPage snapshot 1`] = `
<MyPageWrapper>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <MyPage>
      <styled.div>
        <div
          className="sc-jNMdTA gFRJNv"
        >
          <Header
            line={true}
          >
            <styled.div
              style={
                Object {
                  "borderBottom": "5px solid rgba(236, 47, 22, 1)",
                }
              }
            >
              <div
                className="sc-iktFzd uGybn"
                style={
                  Object {
                    "borderBottom": "5px solid rgba(236, 47, 22, 1)",
                  }
                }
              >
                <styled.div
                  onClick={[Function]}
                >
                  <div
                    className="sc-cBNfnY bDxDMJ"
                    onClick={[Function]}
                  >
                    Gj√∏r No'
                  </div>
                </styled.div>
                <styled.div>
                  <div
                    className="sc-gWHgXt bcglBl"
                  >
                    <styled.div
                      onClick={[Function]}
                    >
                      <div
                        className="sc-hiSbYr hZBClG"
                        onClick={[Function]}
                      >
                        Aktiviteter
                      </div>
                    </styled.div>
                    <styled.div
                      onClick={[Function]}
                    >
                      <div
                        className="sc-hiSbYr hZBClG"
                        onClick={[Function]}
                      >
                        Arrangementer
                      </div>
                    </styled.div>
                  </div>
                </styled.div>
                <styled.div
                  onClick={[Function]}
                >
                  <div
                    className="sc-jJEJSO cgHETT"
                    onClick={[Function]}
                  >
                    Logg inn
                  </div>
                </styled.div>
              </div>
            </styled.div>
          </Header>
          <styled.div>
            <div
              className="sc-cOajty dLNeee"
            >
              <Styled(styled.h1)>
                <h1
                  className="sc-khAkjo sc-iNqMTl kPEFrp gfEEak"
                >
                  Min side
                </h1>
              </Styled(styled.h1)>
              <Styled(styled.div)>
                <div
                  className="sc-AzgDb sc-jeGSBP gzAEGR kUugXk"
                >
                  <Styled(Styled(styled.div))
                    key="Mine aktiviteter"
                    onClick={[Function]}
                  >
                    <div
                      className="sc-lmoMRL sc-iJuUWI sc-nFpLZ hDtEWp hAXfgU dlyIMt"
                      onClick={[Function]}
                    >
                      Mine aktiviteter
                    </div>
                  </Styled(Styled(styled.div))>
                  <Styled(styled.div)
                    key="Oppmeldte aktiviteter"
                    onClick={[Function]}
                  >
                    <div
                      className="sc-lmoMRL sc-eJMQSu hDtEWp fbRiSa"
                      onClick={[Function]}
                    >
                      Oppmeldte aktiviteter
                    </div>
                  </Styled(styled.div)>
                  <Styled(styled.div)
                    key="Favoritter"
                    onClick={[Function]}
                  >
                    <div
                      className="sc-lmoMRL sc-eJMQSu hDtEWp fbRiSa"
                      onClick={[Function]}
                    >
                      Favoritter
                    </div>
                  </Styled(styled.div)>
                </div>
              </Styled(styled.div)>
              <ActivityDashboard
                isLoading={false}
              >
                <styled.div>
                  <div
                    className="sc-giIncl gxyexG"
                  />
                </styled.div>
              </ActivityDashboard>
            </div>
          </styled.div>
        </div>
      </styled.div>
    </MyPage>
  </Provider>
</MyPageWrapper>
`;
